remote-keypair/src/openpgp_card.rs

```diff
@@ -0,0 +1,359 @@
+        // Verify that card has valid signing key
+        self.try_pubkey()?;
+
+        let mut pgp_mut = self.pgp.borrow_mut();
+        let opt = &mut pgp_mut.transaction().map_err(|e| {
+            SignerError::Connection(format!("could not start transaction with card: {e}"))
+        })?;
+        let card_info: OpenpgpCardInfo = opt
+            .try_into()
+            .map_err(|e| SignerError::Connection(format!("could not get card info: {e}")))?;
+
+        // Prompt user for PIN verification if and only if
+        //   * Card indicates PIN is only valid for one PSO:CDS command at a time, or
+        //   * PIN has not yet been entered for the first time.
+        if card_info.pin_cds_valid_once || !*self.pin_verified.borrow() {
+            let mut pin = get_pin_from_user_as_bytes(&card_info, true)?;
+            while opt.verify_pw1_sign(pin.as_bytes()).is_err() {
+                pin = get_pin_from_user_as_bytes(&card_info, false)?;
+            }
+            *self.pin_verified.borrow_mut() = true;
+        }
+
+        // Await user touch confirmation if and only if
+        //   * Card supports touch confirmation, and
+        //   * Touch policy set anything other than "off".
+        if let Some(signing_uif) = card_info.signing_uif {
+            if signing_uif.touch_policy().touch_required() {
+                println!("Awaiting touch confirmation...");
+            }
+        }
+
+        // Delegate message signing to card
+        let hash = Hash::EdDSA(message);
+        let sig = opt
+            .signature_for_hash(hash)
+            .map_err(|e| SignerError::Protocol(format!("card failed to sign message: {e}")))?;
+
+        Ok(Signature::new(&sig[..]))
+    }
+
+    fn is_interactive(&self) -> bool {
+        true
+    }
+}
+
+/// Data struct for convenience.
+#[derive(Debug)]
+struct OpenpgpCardInfo {
+    aid: ApplicationIdentifier,
+    cardholder_name: String,
+    signing_uif: Option<UIF>,
+    pin_cds_valid_once: bool,
+}
+
+impl TryFrom<&mut OpenPgpTransaction<'_>> for OpenpgpCardInfo {
+    type Error = openpgp_card::Error;
+
+    fn try_from(opt: &mut OpenPgpTransaction) -> Result<Self, Self::Error> {
+        let ard = opt.application_related_data()?;
+        Ok(OpenpgpCardInfo {
+            aid: ard.application_id()?,
+            cardholder_name: String::from_utf8_lossy(
+                opt.cardholder_related_data()?.name().get_or_insert(b"null"),
+            )
+            .to_string(),
+            signing_uif: ard.uif_pso_cds()?,
+            pin_cds_valid_once: ard.pw_status_bytes()?.pw1_cds_valid_once(),
+        })
+    }
+}
+
+fn get_pin_from_user_as_bytes(
+    card_info: &OpenpgpCardInfo,
+    first_attempt: bool,
+) -> Result<String, SignerError> {
+    let description = format!(
+        "\
+            Please unlock the card%0A\
+            %0A\
+            Manufacturer: {}%0A\
+            Serial: {:X}%0A\
+            Cardholder: {}\
+            {}\
+        ",
```

Comment:
"\n" is a common escape sequence for the newline character.
While %0A works, it is non-standard and could be confusing.

```suggestion
        "Please unlock the card\n\
         \n\
         Manufacturer: {}\n\
         Serial: {:X}\n\
         Cardholder: {}\
         {}",
```

I think a common Rust syntax convention is to indent multiline string by one space, to align the first character of the string in the first line with the first characters on all the subsequent lines.

Reply:
Sorry, I did not realize those are literal 3 character sequences, rather than escape codes.
Multiline string indentation comment still applies :P